src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+            item->_direction = SyncFileItem::Up;
+            item->_checksumHeader.clear();
+            item->_size = localEntry.size;
+            item->_modtime = localEntry.modtime;
+            item->_type = localEntry.isDirectory ? ItemTypeDirectory : localEntry.isVirtualFile ? ItemTypeVirtualFile : ItemTypeFile;
+            if (!_dirItem || _dirItem->_direction == SyncFileItem::Down) {
+                _childModified = true;
+            }
+
+            auto postProcessLocalNew = [item, localEntry, this]() {
+                if (localEntry.isVirtualFile) {
+                    // Remove the spurious file if it looks like a placeholder file
+                    // (we know placeholder files contain " ")
+                    if (localEntry.size <= 1) {
+                        qCWarning(lcDisco) << "Wiping virtual file without db entry for" << _currentFolder._local + "/" + localEntry.name;
+                        item->_instruction = CSYNC_INSTRUCTION_REMOVE;
+                        item->_direction = SyncFileItem::Down;
+                    } else {
+                        qCWarning(lcDisco) << "Virtual file without db entry for" << _currentFolder._local << localEntry.name
+                                           << "but looks odd, keeping";
+                        item->_instruction = CSYNC_INSTRUCTION_IGNORE;
+                    }
+                }
+            };
+
+            // Check if it is a move
+            OCC::SyncJournalFileRecord base;
+            if (!_discoveryData->_statedb->getFileRecordByInode(localEntry.inode, &base)) {
+                dbError();
+                return;
+            }
+            bool isMove = base.isValid() && base._type == item->_type
+                && ((base._modtime == localEntry.modtime && base._fileSize == localEntry.size)
+                       // Directories and virtual files don't need size/mtime equality
+                       || item->_type == ItemTypeDirectory || localEntry.isVirtualFile);
+
+            if (isMove) {
+                //  The old file must have been deleted.
+                isMove = !QFile::exists(_discoveryData->_localDir + base._path);
+            }
+
+            // Verify the checksum where possible
+            if (isMove && !base._checksumHeader.isEmpty() && item->_type == ItemTypeFile) {
+                if (computeLocalChecksum(base._checksumHeader, _discoveryData->_localDir + path._original, item)) {
+                    qCInfo(lcDisco) << "checking checksum of potential rename " << path._original << item->_checksumHeader << base._checksumHeader;
+                    isMove = item->_checksumHeader == base._checksumHeader;
+                }
+            }
+            auto originalPath = QString::fromUtf8(base._path);
+            if (isMove && _discoveryData->_renamedItems.contains(originalPath))
+                isMove = false;
+
+            //Check local permission if we are allowed to put move the file here
```

Comment:
This move permissions check is a bit surprising at this point as other permissions are checked further down. I expect it has to be done right here though. Hmm, put into function at least?

Reply:
We need to check that the permission are correct before calling findAndCancelDeletedJob, as this has side effect

This allow to convert a move into a remove + download immediately.